File: 30days-of-code/15_linkedlist.cc
#include <iostream>
#include <vector>

using namespace std;	


template<class T>
class Node
{
	public:
	const T data;
	Node<T>* next;

	public:
	explicit Node(const T d)
		: data{d}, next{nullptr}
	{}
};


template<class T>
class MyList {
	public:
	explicit MyList() : tail{nullptr}
	{}

	Node<T>* insert(Node<T>* const head, const T data) {
		Node<T>* new_node = new Node<T>{data};

		if (head == nullptr) {
			tail = new_node;
			return new_node;
		}
		else {
			tail->next = new_node;
			tail = new_node;
			return head;
		}
	}

	void display(Node<T>* const head) const {
		Node<T>* start = head;
		while(start) {
		  cout << start->data << " ";
		  start = start->next;
		}
		cout << endl;
	}

	private:
	Node<T>* tail;
};


template<typename T>
vector<T> read_input() {
	int n{};
	cin >> n;
	const int number_lines{n};

	T item{};
	vector<T> data{};
	for (int i{number_lines}; i>0; --i) {
	    cin >> item;
	    data.push_back(item);
	}

	return data;
}

template<typename T>
void solve(const vector<T> data) {
	Node<T>* head{nullptr};
  	MyList<T> mylist{};

	for (auto item: data) {
	    head = mylist.insert(head, item);
	}

	mylist.display(head);
}


int main() {
	vector<int> data{};
	data = read_input<int>();

	solve(data);
	return 0;
}

File: 30days-of-code/15_linkedlist.py
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None

class Solution:
    def dump(self, head):
        current = head
        while current:
            yield current.data
            current = current.next

    def insert(self, head, data):
        newnode = Node(data)
        if head is None:
            self.tail = newnode
            return newnode
        else:
            self.tail.next = newnode
            self.tail = newnode
            return head


def solve(data):
    mylist = Solution()
    head = None
    for item in data:
        head = mylist.insert(head, item)    

    return mylist.dump(head)


def generate_numbers():
    N = int(input())
    for _ in range(N):
        yield int(input())


def read_numbers():
    N = int(input())
    numbers = []
    for _ in range(N):
        numbers.append(int(input()))
    return numbers


if __name__ == "__main__":
    read_input = generate_numbers

    dumped = solve( read_input() )

    for d in dumped:
        print(d, end=' ')
    print()
